### DISCUSSION TITLE
Public API Services

### DISCUSSION THREAD

**Discussion by Keith-CY** [nervosnet/godwoken_explorer#201] [Ideas]
### Basic JSON RPC: these APIs are from [godwoken-web3](https://github.com/nervosnetwork/godwoken-web3#web3-rpc-modules)

- [ ] net: https://github.com/nervosnetwork/godwoken-web3#net
- [ ] web3: https://github.com/nervosnetwork/godwoken-web3#web3
- [ ] eth: https://github.com/nervosnetwork/godwoken-web3#eth
- [ ] gw: https://github.com/nervosnetwork/godwoken-web3#gw
- [ ] poly: https://github.com/nervosnetwork/godwoken-web3#poly

### Account
- [ ] get ckb/eth balance for a single address
- [ ] get ckb/eth balances for multiple addresses
- [ ] get a list of transactions by address
	- params
		- block range
		- page
		- sort
	- return
		- same as the value returned by JSON RPC
- [ ] <del>get a list of internal transactions by contract address</del>
- [ ] <del>get internal transactions by transaction hash</del>
- [ ] <del>get internal transactions by block range</del>
- [ ] get a list of `erc-20` transfer events by address
	- params
		- block range
		- from address
		- to address
		- contract address
		- page
	- return
		- block number
		- block hash
		- timestamp
		- tx hash
		- tx index
		- contract address
		- from
		- to
		- value
		- token name
		- token symbol
		- token decimal
		- gas used
		- gas price
- [ ] <del>get a list of `erc-721` transfer events by address
- [ ] <del>get historical ckb/eth balances of a single address by block

### Contract
- [ ] get contract abi by address
	- params
		- contract address
	- return
	  - contract abi
- [ ] get contract source code by contract address
	- params
		- contract address
	- return
		- source code
		- abi
		- contract name
		- constructor arguments
		- evm version
- [ ] get erc-20 total supply by contract address
	- params
		- contract address
	- return
		- total supply
- [ ] get erc-20 circulating supply by contract address
	- params
		- contract address
	- return
		- circulating supply
		- updated block number
- [ ] get erc-20 account balance by eoa address and contract address
	- params
		- contract address
		- external account address
	- return
		- balance
- [ ] get token info by contract address
	- params
		- contract address
	- return: most of them are registered in [token submission](https://github.com/nervina-labs/godwoken_explorer/issues/new/choose)
		- contract address
		- token name
		- symbol
		- decimal
		- max supply
		- description
		- website
		- logo uri
		- deployment transaction hash
		- evm version
		- other info

### Transactions
- [ ] get contract execution status
	- params
		- transaction hash
	- return
		- error code: empty if transaction is successful
		- error message

### Blocks
- [ ] <del>get daily average block size</del>
- [ ] <del>get daily block count</del>

### Logs: similar to eth_getLogs with composable topics: https://docs.etherscan.io/api-endpoints/logs
- [ ] get logs by filters

### Tracker
- [ ] get estimation of confirmation time(finalized state)
	- return
		- estimation time in seconds
- [ ] <del>get daily average gas limit</del>
- [ ] <del>get daily average gas used</del>
- [ ] <del>get daily average gas price</del>

### Stats
- [ ] <del>get circulating supply of a token(including bridged and native tokens)</del>

**Comment by Keith-CY** [comment]
APIs should be compatible with https://github.com/nervosnetwork/godwoken-web3

**Comment by Keith-CY** [comment]
The APIs listed above are so primary in DApp development that we are going to deliver them with the alpha of API service (by 05/20)

[godwoken-web3](https://github.com/nervosnetwork/godwoken-web3) is designed to be a primitive backend of DApps, so we can build the service on it and extend its ability with concurrency of elixir.

Any idea about the schema and priority of APIs,  or the structure of public API service?

@zmcNotafraid @Naupio

**Reply by Naupio** [reply]
we suggest provide the GraphQL API Service for used

**Reply by Naupio** [reply]
We designing the API protocol here.
Testing environment: https://api.staging.gwscan.com/graphql

**Comment by Keith-CY** [comment]
From user's feedback, dependent APIs to fetch `deposit records` and `withdrawal records` are requested.

The schema they desire is as follows
```
interface WithdrawalRequest {
  txHash: string;
  blockNumber: number;
  capacity: string;
  lock: Script;
  outPoint: OutPoint;
  type?: Script;
  sudtAmount?: string;
}

interface Script {
  codeHash: string;
  args: string;
  hashType: string;
}

interface OutPoint {
  txHash: string;
  index: string;
}
```

Any idea? @Naupio @zmcNotafraid

**Comment by Keith-CY** [comment]
There're two feedbacks from acceptance team
1. Should we return all values/balances in Bigint instead of decimal for a broadened usage, ref: https://discord.com/channels/956765352514183188/958261584004804650/971397481306935336
2. We've suffered a lot with pagination because we need to specify the page number explicitly. In most cases we specify the page number in the block list, so if we can use a cursor in the transaction list to avoid cumbersome calculation of total count? And thanks to the order of blocks, we can calculate the cursor of block list in front-end which also accelerates the query. For the transaction list, we can provides the closest several cursors for user to specify just like what google does, ref: https://discord.com/channels/956765352514183188/958261584004804650/971399723254706206

**Reply by Keith-CY** [reply]
For the second point, I'm going to adopt graphql in gwscan so the restful api could be intact

**Reply by Naupio** [reply]
> There're two feedbacks from acceptance team
> 
>     1. Should we return all values/balances in Bigint instead of decimal for a broadened usage, ref: https://discord.com/channels/956765352514183188/958261584004804650/971397481306935336
> 
>     2. We've suffered a lot with pagination because we need to specify the page number explicitly. In most cases we specify the page number in the block list, so if we can use a cursor in the transaction list to avoid cumbersome calculation of total count? And thanks to the order of blocks, we can calculate the cursor of block list in front-end which also accelerates the query. For the transaction list, we can provides the closest several cursors for user to specify just like what google does, ref: https://discord.com/channels/956765352514183188/958261584004804650/971399723254706206

For the first point, the account_udt object with `balance` field was original value of shannon, we will change this field type to new custom scalar type which name `Bigint`

**Comment by Keith-CY** [comment]
Account page requires a token_transfer list of a specific address, including `from_address = account` and `to_address = account`
